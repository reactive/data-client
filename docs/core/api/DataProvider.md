---
title: DataProvider - Normalized async data management in React
sidebar_label: <DataProvider />
description: High performance, globally consistent data management in React
---

import Installation from '../shared/\_installation.mdx';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# &lt;DataProvider />

Manages state, providing all context needed to use the hooks. Should be placed as high as possible
in application tree as any usage of the hooks is only possible for components below the provider
in the React tree.

<Installation />

## Props

```typescript
interface ProviderProps {
  children: ReactNode;
  managers?: Manager[];
  initialState?: State<unknown>;
  Controller?: typeof Controller;
  devButton?:
    | 'bottom-right'
    | 'bottom-left'
    | 'top-right'
    | 'top-left'
    | null;
}
```

### initialState: State&lt;unknown\> {#initialState}

```typescript
export interface State<T> {
  readonly entities: {
    readonly [entityKey: string]: { readonly [pk: string]: T } | undefined;
  };
  readonly endpoints: {
    readonly [key: string]: unknown | PK[] | PK | undefined;
  };
  readonly indexes: NormalizedIndex;
  readonly meta: {
    readonly [key: string]: {
      readonly date: number;
      readonly error?: ErrorTypes;
      readonly expiresAt: number;
      readonly prevExpiresAt?: number;
      readonly invalidated?: boolean;
      readonly errorPolicy?: 'hard' | 'soft' | undefined;
    };
  };
  readonly entitiesMeta: {
    readonly [entityKey: string]: {
      readonly [pk: string]: {
        readonly date: number;
        readonly expiresAt: number;
        readonly fetchedAt: number;
      };
    };
  };
  readonly optimistic: (SetResponseAction | OptimisticAction)[];
  readonly lastReset: number;
}
```

Instead of starting with an empty cache, you can provide your own initial state. This can
be useful for testing, or rehydrating the cache state when using server side rendering.

### managers?: Manager[] {#managers}

List of [Manager](./Manager.md)s use. This is the main extensibility point of the provider.

[getDefaultManagers()](./getDefaultManagers.md) can be used to extend the default managers.

Default Production:

```typescript
[new NetworkManager(), new SubscriptionManager(PollingSubscription)];
```

Default Development:

```typescript
[
  new DevToolsManager(),
  new NetworkManager(),
  new SubscriptionManager(PollingSubscription),
];
```

### Controller: typeof Controller {#Controller}

This allows you to extend [Controller](./Controller.md) to provide additional functionality.
This might be useful if you have additional actions you want to dispatch to custom [Managers](./Manager.md)

```tsx
class MyController extends Controller {
  doSomething = () => {
    console.log('hi');
  };
}

const RealApp = (
  <DataProvider Controller={MyController}>
    <App />
  </DataProvider>
);
```

### devButton

<img src="/img/client-logo.svg" style={{float:'right',width:'40px'}} />

In development, a small button will appear that gives easy access to [browser devtools](../getting-started/debugging.md) if
installed. This option configures where it shows up, or if null will disable it altogether.

`'bottom-right' | 'bottom-left' | 'top-right'| 'top-left' | null` = `'bottom-right'`

```tsx title="Disable button"
<DataProvider devButton={null}>
  <App/>
</DataProvider>
```

```tsx title="Place in top right corner"
<DataProvider devButton="top-right">
  <App/>
</DataProvider>
```
