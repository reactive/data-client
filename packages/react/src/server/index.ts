export { default as createPersistedStore } from './createPersistedStore.js';
export * from './getInitialData.js';
export { default as createServerDataComponent } from './createServerDataComponent.js';
export { default as ServerData } from './ServerData.js';
