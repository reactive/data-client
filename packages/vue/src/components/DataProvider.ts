import { Controller as DataController } from '@data-client/core';
import type { State, Manager, GCInterface } from '@data-client/core';
import { defineComponent } from 'vue';

import { provideDataClient } from '../providers/provideDataClient.js';

export interface ProviderProps {
  managers?: Manager[];
  initialState?: State<unknown>;
  Controller?: typeof DataController;
  gcPolicy?: GCInterface;
}

export default defineComponent<ProviderProps>({
  name: 'DataProvider',
  props: ['managers', 'initialState', 'Controller', 'gcPolicy'] as any,
  setup(props, { slots }) {
    provideDataClient({
      Controller: props.Controller,
      gcPolicy: props.gcPolicy,
      initialState: props.initialState,
      managers: props.managers,
    });
    return () => (slots.default ? slots.default() : null);
  },
});
