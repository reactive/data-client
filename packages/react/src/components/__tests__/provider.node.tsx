// eslint-env jest
import React, { version } from 'react';
import { renderToString } from 'react-dom/server';

import DataProvider from '../DataProvider';

describe('<DataProvider />', () => {
  let warnspy: jest.Spied<any>;
  beforeEach(() => {
    warnspy = jest.spyOn(global.console, 'warn').mockImplementation(() => {});
  });
  afterEach(() => {
    warnspy.mockRestore();
  });

  it('should warn users about SSR with DataProvider', async () => {
    const tree = (
      <DataProvider>
        <div>hi</div>
      </DataProvider>
    );
    const LegacyReact = version.startsWith('16') || version.startsWith('17');

    // let our lazy component load
    renderToString(tree);
    await new Promise(resolve => setTimeout(resolve, 0));

    const msg = renderToString(tree);

    expect(warnspy.mock.lastCall).toMatchSnapshot();

    if (!LegacyReact) {
      expect(msg).toMatchInlineSnapshot(
        `"<!--$--><div>hi</div><!--/$--><!--$--><!--/$-->"`,
      );
    } else {
      expect(msg).toMatchInlineSnapshot(`"<div>hi</div>"`);
    }
  });
});
