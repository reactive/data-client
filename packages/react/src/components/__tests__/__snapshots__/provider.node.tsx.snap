// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<DataProvider /> should warn users about SSR with DataProvider 1`] = `
[
  "DataProvider from @data-client/react does not update while doing SSR.
See https://dataclient.io/docs/guides/ssr.",
]
`;
